src/plugin-host/plugin/plugin.ts

```diff
@@ -0,0 +1,139 @@
+import { PluginEventArguments } from "../plugin-events/plugin-event-arguments";
+import { PLUGIN_EVENT } from "../plugin-events/plugin-event-types";
+import { UserScoreChangedPluginEventArguments } from "../plugin-events/event-arguments/user-score-changed-plugin-event-arguments";
+import { PrePostMessagePluginEventArguments } from "../plugin-events/event-arguments/pre-post-message-plugin-event-arguments";
+import { LeaderboardResetPluginEventArguments } from "../plugin-events/event-arguments/leaderboard-reset-plugin-event-arguments";
+import { NoArgumentsPluginEventArguments } from "../plugin-events/event-arguments/no-arguments-plugin-event-arguments";
+import { ChatServices } from "../plugin-chat-services/chat-services";
+import { PluginCommand } from "./plugin-command";
+
+/**
+ * Class defining the interface every plugin should adhere to.
+ */
+export abstract class AbstractPlugin 
+{
+  /**
+   * Semantic identifier of this plugin. 
+   */
+  public Name: string;
+  /**
+   * Version of this plugin.
+   */
+  public Version: string;
+  /**
+   * Boolean indicating the status of this plugin.
+   * Disabled plugins will not receive messages
+   * from its Plugin Host.
+   */
+  public Enabled: boolean;
+  /**
+   * Internal plugin state.
+   */
+  protected Data: any;
+
+  /**
+   * Services for this chat
+   */
+  public Services: () => ChatServices;
+
+  /**
+   * Plugin ID
+   */
+  public PID: () => string;
+
+  /**
+   * Event triggers. Plugins can hook functions to certain Plugin Events.
+   * These plugin events are defined in the PLUGIN_EVENT enumeration.
+   */
+  private pluginEventTriggers: Map<PLUGIN_EVENT, (data: any) => any>;
+
+  /**
+   * Command triggers. Plugins can hook functions to certain commands.
+   * these commands can be set in the constructor of a plugin.
+   */
+  private pluginCommandTriggers: PluginCommand[];
+
+  /**
+   * Create a new Plugin instance.
+   * @param _name Semantic name of this plugin.
+   * @param _version Version of this plugin.
+   * @param _data Any optional data that might require to be stored
+   *              for this plugin.
+   */
+  constructor(_name: string, _version: string, _data: any) 
+  {
+    this.Name = _name;
+    this.Version = _version;
+    this.Data = _data;
+    this.pluginEventTriggers = new Map<PLUGIN_EVENT, (data: any) => any>();
+    this.pluginCommandTriggers = [];
+
+    console.log("+ Loaded plugin: " + this.Name + " (" + this.PID + ")");
+  };
+
+  /* Function overload list */
+  protected subscribeToPluginEvent(_event: PLUGIN_EVENT.PLUGIN_EVENT_PRE_MESSAGE, _eventFn: (_data: PrePostMessagePluginEventArguments) => any): void;
```

Comment:
I'm not entirely sure what these lists are for. How do these work in the overall architecture of the plugins system?

Reply:
These are just non-generic overloads for one function.